packages/x-data-grid/src/hooks/features/virtualization/useGridVirtualScroller.tsx

```diff
@@ -207,6 +204,37 @@ export const useGridVirtualScroller = () => {
   const previousContextScrollPosition = React.useRef(EMPTY_SCROLL_POSITION);
   const previousRowContext = React.useRef(EMPTY_RENDER_CONTEXT);
   const renderContext = useGridSelector(apiRef, gridRenderContextSelector);
+  const focusedVirtualCell = useGridSelector(
+    apiRef,
+    () => {
+      const currentRenderContext = gridRenderContextSelector(apiRef);
+      const focusedCell = gridFocusCellSelector(apiRef);
+      if (!focusedCell) {
+        return null;
+      }
+
+      const rowIndex = currentPage.rows.findIndex((row) => row.id === focusedCell.id);
+      const columnIndex = visibleColumns.findIndex((column) => column.field === focusedCell.field);
+
+      if (rowIndex === -1 || columnIndex === -1) {
+        return null;
+      }
+
+      const isFocusedCellInContext =
+        currentRenderContext.firstRowIndex <= rowIndex &&
+        rowIndex <= currentRenderContext.lastRowIndex;
+
+      if (isFocusedCellInContext) {
+        return null;
+      }
+      return {
+        ...focusedCell,
+        rowIndex,
+        columnIndex,
+      };
+    },
+    objectShallowCompare,
+  );
```

Comment:
Another case of "should be a selector with arguments", but this is low-impact, so you could leave a TODO comment and I'm ok with it.

Reply:
Though now that I re-read, those `findIndex()` calls aren't great. Those are `O(n)` costs that are called everytime the state updates, I think it would be preferable to memoize that logic somehow (either through `createSelectorMemoized` or `React.useMemo`).

Reply:
Good catch. Reduced the lookup cost now, but if you have any ideas on how to make it even better, I'm all ears. Would `rowIdToRowIndex` lookup map make any sense?

Reply:
Pushed another alternative with lookup maps.

Reply:
I think I don't love the lookup maps, they add an `O(n)` cost (plus a bunch of memory I/O) when the rows are updated to build the lookup map. For the use-cases where rows are updated frequently, it could have an effect. I would favor going back to memoization. If you removed `React.useMemo`, I guess selector memoization would be the alternative. Here is an example how I think it could work (some of the dependency selectors are placeholders, can't remember the right names atm):

https://gist.github.com/romgrk/7f7a682578a1c5f2bbcd8b82aef3148e

Reply:
As long as there is at least one `O(n)` operation per setting data (which seems to be the case, looking at e.g. `gridExpandedSortedRowEntriesSelector`, but probably not the only one), adding a global map doesn't need to come at an additional cost, and memory aspect can be negated by using the row object reference as key of the map rather than the id. Things could probably be simplified if `useGridSelector` could take non-selector dependencies as well, similar to `useMemo`.

If we can avoid it, I'm all for it, but now when I looked into the implementation, it can become a bit hairy, unless we also refactor `getVisibleRows`. But that we may need to do anyways though, as I just noticed that `useGridVisibleRows` calls an `O(n)` operation on each render by the looks of it, which is a much bigger issue, since the hook is used in 10+ places.

Reply:
There's currently at least 4 O(n) operations that I found, potentially more, on the example of https://mui.com/x/react-data-grid/#pro-plan:
```
gridSortedRowEntriesSelector 100000
gridExpandedSortedRowEntriesSelector 100000
gridExpandedSortedRowIdsSelector 100000
gridExpandedSortedRowTreeLevelPositionLookupSelector 100000
```
(number shows how many rows the process)

As long as there's at least one of them, adding a lookup map doesn't have to come at an additional cost as long as we can do the operations in parallel. And memory issue can probably be negated by using the row object reference as the map key.

If we can avoid all of them, even better, but maybe not very unlikely? And if so, wouldn't it be better to create the map alongside the other O(n) operations that run an O(n) operation on each focus/rendercontext change?

Reply:
Seems like there are many places that need to calculate a rowIndex that could potentially be optimised with a lookup map.

Reply:
`sortedIds` array could be `rowIdToIndex` map instead. Complexity stays the same, but further optimisations possible in the rest of the grid

Reply:
I don't think I can use `createSelectorMemoized` for this – renderContext will trigger re-evaluation, and we need to re-calculate the `rowIndex`. Unless we start caching focusCell separately in a ref or if we could access previous selector value. However, that makes the code very fragile (what if the sort order changes while the focus cell stays the same? or column order, etc.). Hence I don't currently see a better option than a `lookupMap`, or having `VirtualScroller` re-render on each focus change as it does currently, as then we have the lookup cost only on changing focus, rather than scroll, but that's still less optimal for keyboard navigation in a large grid than a `lookupMap` imho.

The 2nd best option was the first that I submitted – reduce the cost in selector to renderContext size, and calculate real indexes in a `useMemo` for whenever `currentPage.rows` changes.

https://github.com/mui/mui-x/pull/15666/commits/71bffd7ed5b465813f1fbe63b8376e35a61f420a

Reply:
Pushed a quick example of what I'd do – lookupMap creation takes 10ms in dev for 100k rows under page load (probably less under normal conditions). Would take anywhere from 60-100% longer if we used rowId as the map key instead of object reference, and consume much more memory with uuid ids.

Reply:
Forgot to update the selector itself, which can now be improved thanks to the change to useGridVisibleRows. But you get the gist.

Reply:
> This selector needs objectShallowCompare still, since it revalidates every time renderContextSelector changes, however, the resulting derived state doesn't always change when renderContextSelector changes, but we can't provide a stable object reference. Would it be different with the v8 selectors?

The trick would be to create intermediate selectors, so that you can compute e.g. `isFocusedCellInContext` when `renderContext` changes, but the value of `isFocusedCellInContext` would stay the same which wouldn't recompute the final `focusedVirtualCell` selector. I think `columnIndex` would also need to come from an intermediate selector.

You can keep the intermediate selectors in the new file, no need to expose every selector publicly.

Reply:
Would intermediate selectors also have to be memoized or is it enough if the parent selector is? 🤔

Reply:
A selector should be memoized if either of these is true:
 - The selector returns a non-primitive value (object, array, set, map, ...)
 - The selector is computationally expensive (e.g. `O(n)` operation)

So for example the `isFocusedCellInContextSelector` would not need to be memoized, because it returns a boolean which is referentially equal across runs.